packages/ts/file-router/src/runtime/RouterBuilder.ts

```diff
@@ -0,0 +1,133 @@
+import { protectRoutes, type RouteObjectWithAuth } from '@vaadin/hilla-react-auth';
+import type { ReactElement } from 'react';
+import { createBrowserRouter, type RouteObject } from 'react-router-dom';
+import { transformTreeSync } from '../shared/transformTree.js';
+import type { AgnosticRoute } from '../types.js';
+import { toReactRouter } from './toReactRouter.js';
+
+/**
+ * Deeply merges two lists of routes. If the specific path is already present,
+ * the route is merged, otherwise the new routes are added to the list.
+ *
+ * @param a - The first list of routes.
+ * @param b - The second list of routes.
+ */
+function mergeRoutes(a: RouteObject[], b: RouteObject[]): RouteObject[] {
+  return b.reduce(
+    (result, route) => {
+      const existingRoute = result.find((r) => r.path === route.path);
+      if (existingRoute) {
+        Object.assign(existingRoute, route);
+        existingRoute.children = existingRoute.children
+          ? mergeRoutes(existingRoute.children, route.children ?? [])
+          : route.children;
+      } else {
+        result.push(route);
+      }
+      return result;
+    },
+    [...a],
+  );
+}
```

Comment:
Not a real change request, but in general creating a `Map` out of the first array would avoid `find`ing all the time. Given the small size of involved arrays, though, I'm not sure that this compensates the map creation cost.

Reply:
Yup, Copilot suggested the same 😅 Though I'm not sure if it worth it: usually, there is not so many routes in the app to really care about it, and especially it's not so many routes as a child to some parent route. I expect it to be like 1-10 which gives neglectable performance impact IMO.